substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,330 @@
+#[derive(Debug, Clone)]
+pub struct SpanMessage {
+	pub id: Id,
+	pub parent_id: Option<Id>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub start_time: DateTime<Utc>,
+	pub overall_time: chrono::Duration,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// Finished Trace Data Format. Ready for insertion into a relational database.
+#[derive(Debug)]
+pub struct Traces {
+	block_num: u32,
+	hash: Vec<u8>,
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+impl Traces {
+	pub fn new(block_num: u32, hash: Vec<u8>, events: Vec<EventMessage>, spans: Vec<SpanMessage>) -> Self {
+		Traces { block_num, hash, spans, events }
+	}
+
+	/// Get the hash these spans come from.
+	pub fn hash(&self) -> Vec<u8> {
+		self.hash.clone()
+	}
+
+	/// Get the block number of the block these spans come from.
+	pub fn block_num(&self) -> u32 {
+		self.block_num
+	}
+}
+
+#[derive(Debug)]
+pub struct SpanEvents {
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+/// Collects traces and filters based on target.
+/// The Subscriber implementation is blocking. It uses Mutex primitives to coalesce traces before
+/// sending them to the appropriate actor.
+/// Therefore, one must be careful not to block the async executor when adding tracing spans
+/// using this subscriber implementation anywhere inside an async context in substrate-archive.
+pub struct TraceHandler {
+	/// the Block Number this trace set is from
+	block_num: u32,
+	/// Hash of the block for this trace set
+	hash: BlockHash,
+	span_events: Arc<Mutex<SpanEvents>>,
+	targets: Vec<(String, Level)>,
+	counter: AtomicU64,
+	current_span: CurrentSpan,
+}
+
+impl TraceHandler {
+	pub fn new(targets: &str, block_num: u32, hash: BlockHash, span_events: Arc<Mutex<SpanEvents>>) -> Self {
+		let targets = targets.split(',').map(|s| parse_target(s)).collect();
+		// must start indexing from 1 otherwise `tracing` panics
+		let counter = AtomicU64::new(1);
+		Self { targets, counter, span_events, block_num, hash, current_span: Default::default() }
+	}
+
+	/// Formats an event as an `EventMessage` and sends it to the TracingActor.
+	fn gather_event(&self, event: &Event<'_>, time: DateTime<Utc>) -> Result<()> {
+		let meta = event.metadata();
+		let mut values = TraceData::default();
+		event.record(&mut values);
+		let parent_id =
+			event.parent().cloned().or_else(|| self.current_span.id()).ok_or(TracingError::ParentNotFound)?;
+
+		let (target, name) = if meta.name() == WASM_TRACE_IDENTIFIER {
+			match (values.0.remove(WASM_NAME_KEY), values.0.remove(WASM_TARGET_KEY)) {
+				(Some(name), Some(target)) => (name.to_string(), target.to_string()),
+				(Some(name), None) => (name.to_string(), meta.name().to_string()),
+				(None, Some(target)) => (meta.name().to_string(), target.to_string()),
+				(None, None) => (meta.name().to_string(), meta.target().to_string()),
```

Comment:
What are these cases coming from? Is it common that substrate logs without name/target?

Reply:
If our traces are coming from the compiled wasm binary, then the native `name`/`target` will just be `wasm_tracing`, while the real `name`/`target` is stored in the values field of the actual trace. Due to the wasm/native barrier, native doesn't know anything about the trace except that it's a trace that's from wasm.

Reply:
Right, so `(None, None)` is a trace from the runtime? Maybe best comment on each case to keep things clear for our future selves.

Reply:
Changed the block to: https://github.com/paritytech/substrate-archive/pull/134/commits/d4ef899f02fbc2ebd479c99687e4d2849e5711ed

```rust
    let name = values.0.remove(WASM_NAME_KEY).map(|t| t.to_string()).unwrap_or(meta.name().to_string());
    let target = values.0.remove(WASM_TARGET_KEY).map(|t| t.to_string()).unwrap_or(meta.target().to_string());
```